rust/cubesql/cubesql/src/compile/rewrite/rules/split/dates.rs

```diff
@@ -136,19 +157,86 @@ impl SplitRules {
-                    ));
-                    subst.insert(output_granularity_var, output_granularity);
-                    return true;
+                let date_part: DatePartToken = if let Ok(date_part) = date_part.parse() {
+                    date_part
+                } else {
+                    return false;
+                };
+                if !matches!(
+                    date_part,
+                    // Julian sates are very similar to epoch, so we handle it here as well
+                    DatePartToken::Special(SpecialTimeUnitToken::Epoch)
+                        | DatePartToken::Special(SpecialTimeUnitToken::Julian)
+                ) {
+                    return false;
                 }
+
+                let new_date_part = egraph.add(LogicalPlanLanguage::LiteralExprValue(
+                    LiteralExprValue(ScalarValue::Utf8(Some(date_part.as_str().to_string()))),
+                ));
+                subst.insert(new_date_part_var, new_date_part);
+
+                return true;
+            }
+            false
+        }
+    }
+
+    fn transform_date_part(
+        &self,
+        date_part_var: &str,
+        new_date_part_var: &str,
+        new_trunc_granularity_var: &str,
+    ) -> impl Fn(
+        bool,
+        &mut egg::EGraph<LogicalPlanLanguage, LogicalPlanAnalysis>,
+        &mut egg::Subst,
+    ) -> bool
+           + Sync
+           + Send
+           + Clone {
+        let date_part_var = var!(date_part_var);
+        let new_date_part_var = var!(new_date_part_var);
+        let new_trunc_granularity_var = var!(new_trunc_granularity_var);
+        move |_, egraph, subst| {
+            // We are going to split single DatePart to DatePart(DateTrunc)
+            // Parse token as DatePartToken and use DatePartToken::delta_for_trunc to determine appropriate argument for DateTrunc
+            // Also there are several different date parts with same granularity, but different semantic
+            // date_part(dow) and date_part(doy) should both use date_trunc(day), but they return different ranges of values
+            // So we keep original part for date_part after split
+
+            if let Some(ConstantFolding::Scalar(ScalarValue::Utf8(Some(date_part)))) =
+                &egraph[subst[date_part_var]].data.constant
+            {
+                let date_part: DatePartToken = if let Ok(date_part) = date_part.parse() {
+                    date_part
+                } else {
+                    return false;
+                };
+                let trunc_granularity = if let Some(token) = date_part.delta_for_trunc() {
+                    token
+                } else {
+                    return false;
+                };
+
+                let new_date_part = egraph.add(LogicalPlanLanguage::LiteralExprValue(
+                    LiteralExprValue(ScalarValue::Utf8(Some(date_part.as_str().to_string()))),
+                ));
```

Comment:
Same as above; we're copying a constant-folded scalar, so we don't need `?new_date_part` at all.